vllm/logger.py

```diff
@@ -1,72 +1,93 @@
+VLLM_LOGGING_CONFIG_PATH = os.getenv("VLLM_LOGGING_CONFIG_PATH")
 
 _FORMAT = "%(levelname)s %(asctime)s %(filename)s:%(lineno)d] %(message)s"
 _DATE_FORMAT = "%m-%d %H:%M:%S"
 
-
-class NewLineFormatter(logging.Formatter):
-    """Adds logging prefix to newlines to align multi-line messages."""
-
-    def __init__(self, fmt, datefmt=None):
-        logging.Formatter.__init__(self, fmt, datefmt)
-
-    def format(self, record):
-        msg = logging.Formatter.format(self, record)
-        if record.message != "":
-            parts = msg.split(record.message)
-            msg = msg.replace("\n", "\r\n" + parts[0])
-        return msg
-
-
-_root_logger = logging.getLogger("vllm")
-_default_handler: Optional[logging.Handler] = None
-
-
-def _setup_logger():
-    _root_logger.setLevel(logging.DEBUG)
-    global _default_handler
-    if _default_handler is None:
-        _default_handler = logging.StreamHandler(sys.stdout)
-        _default_handler.flush = sys.stdout.flush  # type: ignore
-        _default_handler.setLevel(logging.INFO)
-        _root_logger.addHandler(_default_handler)
-    fmt = NewLineFormatter(_FORMAT, datefmt=_DATE_FORMAT)
-    _default_handler.setFormatter(fmt)
-    # Setting this will avoid the message
-    # being propagated to the parent logger.
-    _root_logger.propagate = False
+DEFAULT_LOGGING_CONFIG = {
+    "formatters": {
+        "vllm": {
+            "class": "vllm.logging.NewLineFormatter",
+            "datefmt": _DATE_FORMAT,
+            "format": _FORMAT,
+        },
+    },
+    "handlers": {
+        "vllm": {
+            "class": "logging.StreamHandler",
+            "formatter": "vllm",
+            "level": "INFO",
+            "stream": "ext://sys.stdout",
+        },
+    },
+    "loggers": {
+        "vllm": {
+            "handlers": ["vllm"],
+            "level": "DEBUG",
+            "propagate": False,
+        },
+    },
+    "version": 1,
+}
+
+
+def _configure_vllm_root_logger() -> None:
+    if VLLM_CONFIGURE_LOGGING:
+        logging_config: Dict = DEFAULT_LOGGING_CONFIG
+
+    if VLLM_LOGGING_CONFIG_PATH:
+        if not path.exists(VLLM_LOGGING_CONFIG_PATH):
+            raise RuntimeError(
+                "Could not load logging config. File does not exist:"
+                f" {VLLM_LOGGING_CONFIG_PATH}")
+        with open(VLLM_LOGGING_CONFIG_PATH, encoding="utf-8",
+                  mode="r") as file:
+            custom_config = json.loads(file.read())
+
+        if not isinstance(custom_config, dict):
+            raise ValueError("Invalid logging config. Expected Dict, got"
+                             f" {type(custom_config).__name__}.")
+        logging_config = custom_config
+
+    if logging_config:
+        dictConfig(logging_config)
+
+
+def _configure_vllm_logger(logger: Logger) -> None:
+    # Use the same settings as for root logger
+    _root_logger = logging.getLogger("vllm")
+    default_log_level = os.getenv("LOG_LEVEL", _root_logger.level)
```

Comment:
There is a change in behavior here! Previously the fallback value was `DEBUG` instead of the log level of the root logger. It seemed that if someone is allowing vllm to configure the logger, they'd want to match the root logger and not arbitrarily apply `DEBUG`. Behavior is only changed from current implementation if someone has reconfigured `_root_logger.level` manually. Otherwise since the root logger has a default level of `DEBUG`, so too should other loggers.

Reply:
(same question as below. If we do propagate=True, isn't this automatically handled?)

Reply:
I've been kind of confused about why this is setup this way also, but tried to defer to the existing implementation.

If `propagate=True` (as it is by default), then this isn't needed. But that relies on all loggers being named along the lines of `vllm.*`. I wondered if this behavior was here to allow for configuring loggers in other namespaces, but that seems a bit funky too.

Here's a basic demo of the behavior is `init_logger` is just an alias for `logging.getLogger`:

```python
from vllm.logger import init_logger

root_logger = init_logger("vllm")
other_logger = init_logger("vllm.other")
unrelated_logger = init_logger("unrelated")

root_logger.info("Root logger")
other_logger.info("Other logger")
unrelated_logger.info("Unrelated")


# [vllm] INFO 05-01 10:37:54 logger_demo.py:7] Root logger
# [vllm.other] INFO 05-01 10:37:54 logger_demo.py:8] Other logger
# unrelated logger has no handlers so is silent
```

Seems good to me. I'm happy to simplify this 👍🏻

Reply:
For the record, I confirmed that all existing usage of `init_logger` would fall into the `vllm.*` namespace. All usage is of the form

```python
# git grep '=.*init_logger' | sed -E 's/^[^:]+: *//g' | sort | uniq -c

logger = init_logger(__name__)
```

Reply:
yeah I think it is a normal practice for libs iiuc (usually packages are all subpackages of the top level init)

Reply:
Changes to simplify things by better utilizing the propagation behavior of loggers have been made 🎉